packages/flutter/lib/src/cupertino/theme.dart

```diff
@@ -303,6 +305,80 @@ class CupertinoThemeData extends NoDefaultCupertinoThemeData with Diagnosticable
   }
 }
 
+class _CupertinoDynamicThemeData with Diagnosticable implements CupertinoThemeData {
+  _CupertinoDynamicThemeData({
+    required CupertinoThemeData baseThemeData,
+    required BuildContext context,
+  }) : _themeData = baseThemeData,
+       _context = context;
+
+  @override
+  Brightness? get brightness => _themeData.brightness;
+
+  @override
+  _CupertinoThemeDefaults get _defaults => _themeData._defaults;
+  final CupertinoThemeData _themeData;
+  final BuildContext _context;
```

Comment:
I tried to put some asserts around the `_context` getter to make sure it's only accessed during persistent callback phase, but that broke some of our tests (those retrieve the theme data and put them in `expect`).